ui/app/src/model/datasource-api.ts

```diff
@@ -11,45 +11,203 @@
+    return fetchGlobalDatasourceList(selector.kind, selector.name ? undefined : true, selector.name).then((list) => {
+      // hopefully it should return at most one element
+      if (list[0] !== undefined) {
+        return {
+          resource: list[0],
+          proxyUrl: getProxyUrl(list[0]),
+        };
+      }
+    });
+  }
+
+  listDatasources(project: string, pluginKind: string | undefined): Promise<Datasource[]> {
+    return fetchDatasourceList(project, pluginKind);
+  }
+
+  listGlobalDatasources(pluginKind: string | undefined): Promise<GlobalDatasource[]> {
+    return fetchGlobalDatasourceList(pluginKind);
+  }
+}
+
+class Cache {
+  private datasources: LRUCache<string, Datasource>;
+  private emptyDatasources: LRUCache<string, boolean>;
+  private globalDatasources: LRUCache<string, GlobalDatasource>;
+
+  constructor() {
+    const option = { ttl: 5 * 60 * 1000, ttlAutopurge: true };
+    this.globalDatasources = new LRUCache<string, GlobalDatasource>(option);
+    this.datasources = new LRUCache<string, Datasource>(option);
+    this.emptyDatasources = new LRUCache<string, boolean>(option);
+  }
+
+  setDatasources(list: Datasource[]) {
+    for (const dts of list) {
+      this.setDatasource(dts);
+    }
+  }
+
+  setDatasource(dts: Datasource) {
+    const kind = dts.spec.plugin.kind;
+    const project = dts.metadata.project;
+    if (dts.spec.default) {
+      // in case it's the default datasource for the given kind, we store it twice
+      // because we might get the datasource with the name or because we want the default one.
+      this.datasources.set(this.generateKey({ kind: kind }, project), dts);
+    }
+    this.datasources.set(this.generateKey({ kind: kind, name: dts.metadata.name }, project), dts);
+  }
+
+  setUndefinedDatasource(project: string, selector: DatasourceSelector) {
+    this.emptyDatasources.set(this.generateKey(selector, project), true);
+  }
+
+  getDatasource(project: string, selector: DatasourceSelector) {
+    const key = this.generateKey(selector, project);
+    const resource = this.datasources.get(this.generateKey(selector, project));
+    let keyExist = true;
+    if (resource === undefined) {
+      keyExist = this.emptyDatasources.has(key);
+    }
+    return { resource: resource, keyExist: keyExist };
+  }
+
+  setGlobalDatasources(list: GlobalDatasource[]) {
+    for (const dts of list) {
+      this.setGlobalDatasource(dts);
+    }
+  }
+
+  setGlobalDatasource(dts: GlobalDatasource) {
+    const kind = dts.spec.plugin.kind;
+    if (dts.spec.default) {
+      // in case it's the default datasource for the given kind, we store it twice
+      // because we might get the datasource with the name or because we want the default one.
+      this.globalDatasources.set(this.generateKey({ kind: kind }), dts);
+    }
+    this.globalDatasources.set(this.generateKey({ kind: kind, name: dts.metadata.name }), dts);
+  }
+
+  setUndefinedGlobalDatasource(selector: DatasourceSelector) {
```

Comment:
I'm not 100% sure it would help here but we've started using [immer](https://immerjs.github.io/immer/) a lot in panels-plugin ([example](https://github.com/perses/perses/blob/main/ui/panels-plugin/src/plugins/gauge-chart/GaugeChartOptionsEditorSettings.tsx)) for visual editing. It could be worth trying it out in the app too in the future

Reply:
mmm that's interesting indeed. Not sure as well it will work here